src/concrete/ml/torch/lora.py

```diff
@@ -31,49 +54,96 @@ class LoraTraining(torch.nn.Module):
+                raise ValueError(
+                    "When no loss_fn is provided, the model's forward method"
+                    "must accept a 'labels' parameter"
+                )
 
-        self.replace_layers_with_custom(self.inference_model, n_layers_to_skip)
+        # Assert that the model contains LoRA layers
+        self.assert_has_lora_layers(model)
 
-        self.optimizer = None
-        self.lr_scheduler = None
-        self.loss_fn = None
-        self.gradient_accumulation_steps = 1
-        self.max_grad_norm = None
+        self.inference_model = model
+        self.replace_layers_with_custom(self.inference_model, n_layers_to_skip_for_backprop)
 
         self.calibrate = False
-        self.run_optimizer = False
+        self.loss_fn = loss_fn
+        self.loss_scaling_factor = 1.0
+
+    def set_loss_scaling_factor(self, loss_scaling_factor: float):
+        """Set the loss scaling factor for gradient accumulation.
+
+        Args:
+            loss_scaling_factor (float): The factor to scale the loss by.
+        """
+        self.loss_scaling_factor = loss_scaling_factor
 
     @staticmethod
-    def replace_layers_with_custom(model: torch.nn.Module, n_layers_to_skip: int):
-        """Replace linear layers with custom ones.
+    def assert_has_lora_layers(model):
+        """Assert that the model contains LoRA layers.
+
+        Args:
+            model (torch.nn.Module): The model to check for LoRA layers.
+
+        Raises:
+            ValueError: If the model does not contain any LoRA layers.
+        """
+
+        def is_lora_module(module):
+            # Check for common LoRA attributes with case-insensitive matching
+            lora_attributes = ["lora_a", "lora_b", "lora_dropout"]
+            return any(
+                hasattr(module, attr)
+                or hasattr(module, attr.lower())
+                or hasattr(module, attr.upper())
+                for attr in lora_attributes
+            )
 
-        This method replaces eligible linear layers in the model with custom layers
-        that are compatible with the LoRA training procedure.
+        has_lora = any(is_lora_module(module) for module in model.modules())
+
+        if not has_lora:
+            raise ValueError("The model does not contain any detectable LoRA layers.")
+
+        print("LoRA layers detected in the model.")
+
+    @staticmethod
+    def replace_layers_with_custom(model: nn.Module, n_layers_to_skip_for_backprop: int) -> None:
+        """Replace linear layers with custom ones.
 
         Args:
-            model (torch.nn.Module): The model to replace layers in.
-            n_layers_to_skip (int): Number of layers to skip.
+            model (nn.Module): The model to replace layers in.
+            n_layers_to_skip_for_backprop (int): Number of initial linear layers to keep as standard
+                layers. Since the first layer doesn't need backpropagation (no previous layer to
```

Comment:
maybe you should change the signature of the function, since you mentioned default to 1

`n_layers_to_skip_for_backprop: int = 1`

Reply:
I suggest: 

n_layers_to_skip_for_backprop (int): Determines how many of the first linear layers are excluded from backpropagation. This is typically set to 1 because the first layer only transforms the input data and does not depend on previous layers for gradient updates. By skipping this layer, we save unnecessary computations. Defaults to 1.

Reply:
Maybe explain why we replace with custom linear layers.
(attach the forward_module and backward_module...)

Reply:
I will remove the default to one here. It's default to one in th LoraTrainer

Reply:
We will have to update the documentation. The definition if this variable is already quite complex

Reply:
alright